hphp/runtime/vm/debug/perf-jitdump.cpp

```diff
@@ -0,0 +1,287 @@
+/*
+   +----------------------------------------------------------------------+
+   | HipHop for PHP                                                       |
+   +----------------------------------------------------------------------+
+   | Copyright (c) 2010-2016 Facebook, Inc. (http://www.facebook.com)     |
+   +----------------------------------------------------------------------+
+   | This source file is subject to version 3.01 of the PHP license,      |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available through the world-wide-web at the following url:           |
+   | http://www.php.net/license/3_01.txt                                  |
+   | If you did not receive a copy of the PHP license and are unable to   |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@php.net so we can mail you a copy immediately.               |
+   +----------------------------------------------------------------------+
+
+ * perf-jitdump.c: perf agent to dump generated code into jit-<pid>.dump
+ *
+ * Adapted from the Oprofile code in opagent.c
+ * Copyright 2007 OProfile authors
+ * Jens Wilke
+ * Daniel Hansel
+ * Copyright IBM Corporation 2007
+ */
+
+#include "hphp/runtime/vm/debug/debug.h"
+#include "hphp/runtime/base/runtime-option.h"
+#include "hphp/runtime/vm/debug/oprof-jitdump.h"
+
+#include <folly/portability/Syscall.h>
+#include <folly/portability/Time.h>
+#include <sys/types.h>
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+#include <errno.h>
+#include <boost/filesystem/operations.hpp>
+#include <boost/filesystem/path.hpp>
+
+using namespace HPHP::jit;
+
+const char padding_bytes[7] = {'\0', '\0', '\0', '\0', '\0', '\0', '\0'};
+
+namespace HPHP {
+namespace Debug {
+
+static int getEMachine(struct jitheader *hdr)  {
+  char id[16];
+  int fd;
+  struct {
+    uint16_t e_type;
+    uint16_t e_machine;
+  } info;
+  
+  fd = open("/proc/self/exe", O_RDONLY);
+  if (fd == -1)  return -1;
+  read(fd, id, sizeof(id));  
+  
+  /* check ELF signature */
+  if (id[0] != 0x7f || id[1] != 'E' || id[2] != 'L' || id[3] != 'F') {
+    close(fd);
+    return -1;
+  }
+  
+  read(fd, &info, sizeof(info));
+  if (info.e_machine < 0)  {
+    hdr->elf_mach = 0; /* ELF EM_NONE */
+  }
+  close(fd);
+  return 0;
+}
+
+static int use_arch_timestamp;
+
+static inline uint64_t getArchTimestamp(void)  {
+#if defined(__i386__) || defined(__x86_64__)
+  unsigned int low, high;
+
+  asm volatile("rdtsc" : "=a" (low), "=d" (high));
+  return low | ((uint64_t)high) << 32;
+#else
+  return 0;
+#endif
+}
+
+/*On Win32 and linux CLOCK_MONOTONIC is available */
+#if !defined(CLOCK_MONOTONIC) && !defined(__APPLE__)
+#define CLOCK_MONOTONIC 1
```

Comment:
Hi Orvid,

If it defined on FB infrastructure then I can remove the `#define`  Please
advise and I can update it today.

Thanks,
sathvik

On Sat, May 7, 2016 at 11:31 AM, Orvid King <notifications@github.com>
wrote:

> In hphp/runtime/vm/debug/perf-jitdump.cpp
> <https://github.com/facebook/hhvm/pull/6926#discussion_r62420692>:
>
> > +static int use_arch_timestamp;
> > +
> > +static inline uint64_t getArchTimestamp(void)  {
> > +#if defined(__i386__) || defined(__x86_64__)
> > +  unsigned int low, high;
> > +
> > +  asm volatile("rdtsc" : "=a" (low), "=d" (high));
> > +  return low | ((uint64_t)high) << 32;
> > +#else
> > +  return 0;
> > +#endif
> > +}
> > +
> > +/*On Win32 and linux CLOCK_MONOTONIC is available */
> > +#if !defined(CLOCK_MONOTONIC) && !defined(__APPLE__)
> > +#define CLOCK_MONOTONIC 1
>
> It should be getting defined when you include on Linux as well, is it not?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/facebook/hhvm/pull/6926/files/2977590b3713d9c29d09e74c56e3da321597ecc0#r62420692>
>